index.js

```diff
@@ -585,20 +627,73 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptedKey) {
     const encryptedVault = this.store.getState().vault;
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    let vault = null;
+    if (encryptedVault && encryptedVault.includes(VAULT_SEPARATOR)) {
+      const [, salt] = encryptedVault.split(VAULT_SEPARATOR);
+
+      if (password) {
+        this.encryptedKey = this._generateEncryptedKey(password, salt);
+      } else if (encryptedKey) {
+        this.encryptedKey = encryptedKey;
+      } else {
+        throw new Error(
+          'No way to decrypt a salted vault without a password or encrypted key',
+        );
+      }
+
+      vault = await this.encryptor.decrypt(this.encryptedKey, encryptedVault);
+    } else {
+      vault = await this.encryptor.decrypt(password, encryptedVault);
+    }
+
     await Promise.all(vault.map(this._restoreKeyring.bind(this)));
+
     await this._updateMemStoreKeyrings();
+
+    // MV3: If we're provided a password, we should persist keyrings
+    // so that we can either (1) migrate or (2) create a new salt
+    if (password) {
+      await this.persistAllKeyrings(password);
+    }
+
     return this.keyrings;
   }
 
+  // MV3:  Generates the encrypted key
+  _generateEncryptedKey(password, salt) {
+    // TODO: How complicaated do we want this to be?
+    return window.btoa(password + salt);
```

Comment:
(must) We want it to be a hash of the two, so that plaintext password is not recoverable. 
(could) Starting with the salt might be preferred for some hashing functions IIRC